include/directx/d3dx12.h

```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
     D3D12_NODE_MASK m_Desc;
 };
 
+//------------------------------------------------------------------------------------------------
+// Implementation for the new CheckFeatureSupport API
+
+#ifdef FEATURE_SUPPORT_USE_VARIANT
+#if (__cplusplus >= 201703L || _MSVC_LANG >= 201703L)
+#define FEATURE_SUPPORT_VARIANT_OPT_IN
+#include <variant>
+#else
+static_assert(false, "std::variant is only supported on compilers with C++17 enabled")
+#endif
+#endif
+
+// Macro to set up a getter function for each entry in feature support data
+// The getter function will have the same name as the feature option name
+#define FEATURE_SUPPORT_GET(RETTYPE,FEATURE,OPTION) \
+RETTYPE OPTION() const \
+{ \
+    return FEATURE.OPTION; \
+}
+
+// Macro to set up a getter function for each entry in feature support data
+// Also specifies the name for the function which can be different from the feature name
+#define FEATURE_SUPPORT_GET_NAME(RETTYPE,FEATURE,OPTION,NAME) \
+RETTYPE NAME() const \
+{\
+    return FEATURE.OPTION; \
+}
+
+// Macro to set up a getter function for feature data indexed by the graphics node ID
+// The default parameter is 0, or the first availabe graphics device node
+#define FEATURE_SUPPORT_GET_NODE_INDEXED(RETTYPE,FEATURE,OPTION) \
+RETTYPE OPTION(UINT NodeIndex=0) const \
+{\
+    return FEATURE[NodeIndex].OPTION; \
+}
+
+// Macro to set up a getter function for feature data indexed by NodeIndex
+// Allows a custom name for the getter function
+#define FEATURE_SUPPORT_GET_NODE_INDEXED_NAME(RETTYPE,FEATURE,OPTION,NAME) \
+RETTYPE NAME(UINT NodeIndex=0) const \
+{\
+    return FEATURE[NodeIndex].OPTION; \
+}
+
+// Macro to initialize a member feature data struct corresponding to the specified feature
+// Will stop the initialization process and return an error code upon failure
+#define INITIALIZE_MEMBER_DATA_CHECKED(FEATURE, MEMBER) \
+if (FAILED(m_hStatus = m_pDevice->CheckFeatureSupport(FEATURE, &MEMBER, sizeof(MEMBER)))) { \
+    return m_hStatus; \
+}
+
+// Macro to initialize a member feature data and report on failure
+// Does not stop the initialization process by itself
+#define INITIALIZE_FAILED(FEATURE, MEMBER) FAILED(m_pDevice->CheckFeatureSupport(FEATURE, &MEMBER, sizeof(MEMBER)))
+
+class CD3DX12FeatureSupport
+{
+public:
+    CD3DX12FeatureSupport()

```

Comment:
Continuation of my comment from preview PR:   I propose putting a raw declaration of all the member functions at the top of the class here.  So users can quickly see all the things they can do.  They can scroll down to see implementations if they want (and you can keep the macros there).

Reply:
Agreed. Will do that in the next commit!

Reply:
It seems C++ doesn't allow me to declare the function then define the same function inside the class declaration scope. I'll have to do the definition outside the scope (moving the implementation out and scope them with CD3DX12FeatureSupport). This is a little bit more complicated than I expect. Since I'm running a little short on time, I'll perform this update when I finalize the PR. Sorry!

Reply:
Resolved. I've separated declarations from definitions and moved definitions to the bottom.